### ISSUE TITLE
Can LSTM models translated using armnn work with functions defined in ACL?

### ISSUE THREAD

**Issue by Pradeepcbk** [ARM-software/armnn#331]
Hello All,

I have a task of performing regression learning on a target board (ARM Cortex A53 processor).

I have completed the tasks of data and feature extraction.

Now, i am analyzing the performances of different learning algorithms (GRUs and LSTM) in tensorflow.

However, my concern is that not all models has the software supports to be deployed on the target board.

Out of the software libraries privided by ARM for ML models, i found out that only ARM NN on top of ARM compute library (ACL) is the way to go. 

Although ACL has the function defnitions needed for the basic defnition of LSTM, in the description of the compute library - it specifies that it supports only SVM (Binary Classification) and CNN models.

Can anyone confirm if the LSTM models translated by ARM NN works with ACL?

Thanks in advance

**Comment by MatthewARM** [comment]
Yes, we've used Mozilla Deep Speech via Arm NN over Compute Library. That uses LSTM and works fine.

**Comment by MatthewARM** [comment]
After thinking about this for a bit, I'm now not sure that the tflite version of Mozilla DeepSpeech actually uses LSTM.

However, LSTM is used by the Android NNAPI and is extensively tested with ArmnNN

**Comment by kejriwalnishant1990** [comment]
@MatthewARM : Does ARM NN has support for tflite/tensorflow weights parser of LSTM layers?

**Comment by Pradeepcbk** [comment]
@MatthewARM Thank you for the response. Is GRU and Random Forest also supported?

Is there any example blogs for the implementations of Android NNAPI and Mozilla DeepSpeech?

**Comment by MatthewARM** [comment]
@kejriwalnishant1990 : No, I've checked and we don't have LSTM support in the parsers, only in the core APU. It shouldn't be too hard to add, given a test model.

@Pradeepcbk We don't have GRU, only LSTM. We don't have Random Forest either.

The Android NNAPI is the neural networks API on Android. Deployment of Arm NN to support the Android NNAPI is typically done by phone manufacturers only. There is more information at https://github.com/ARM-software/android-nn-driver

For Mozilla DeepSpeech, we follow the instructions at https://github.com/mozilla/DeepSpeech/blob/master/doc/TRAINING.rst#exporting-a-model-for-tflite and we test that the exported model loads and runs fine in Arm NN. We've not built a complete test program around it to decode speech files, and I don't know how hard that would be to do.

Hope that helps,
Matthew

**Comment by Pradeepcbk** [comment]
@MatthewARM  Thanks again for the response.

Is there a requirement that the OS has to be Mac OS or linux, or is it possible even with Windows?

**Comment by Pradeepcbk** [comment]
@MatthewARM , Thank you for the response.

With this conversations, i believe that an LSTM model can actually deployed on a ARM Cortex-a53 CPU. I will try that. 

For using the ARMNN or ACL libraries - is there a requirement to use a specific OS environment? Will it work with Windows 10?

**Comment by MatthewARM** [comment]
@rhughesgeomerics has got it working on Windows 10 in the past, I'll ask him to comment.

**Comment by rhughesgeomerics** [comment]
Hi @Pradeepcbk,

The ARMNN libraries are compatible with Windows 10 (I use Windows 10 as my development environment and regularly maintain ARMNN to keep compatibility with Windows 10).

I used to maintain an ACL branch that worked on Windows 10 (using an Intel-provided conversion header from NEON -> SSE intrinsics), but I haven't worked on ACL for at least a year and so this would likely need some effort to get it working again.

Would you be running Windows 10 on the Cortex-A53?

Thanks,
Rob

**Comment by Pradeepcbk** [comment]
@rhughesgeomerics Thank you for the response.

I will be running a customized OS on the processor.

My question was in regards to the development environment. From your response, i believe that it is possible to develop and generate the hex file in windows environment using ARMNN and ACL before flashing it on the Cortex-A53. Please correct me if i am wrong.

Is there an example of the complete process of Tensorflowlite --> ARMNN --> ACL --> Flashed on the target board available anywhere. This will allow me to test the end to end system with a very simple LSTM model once before going ahead with a complicated one.

Thanks in advance.

**Comment by MatthewARM** [comment]
Hi @Pradeepcbk I'm afraid we don't have a complete end-to-end example of what you're trying to do.

Arm NN is a flexible toolkit but at its heart is a neural network runtime engine. You need to get your network into that engine, running on your target board.

There are several ways to do this, but I think the following steps may work for your needs:
* On your Windows 10 development host, build `armnnConverter` with the tflite parser enabled.
* Use armnnConverter to convert the .tflite file to .armnn file.
* Cross-compile your application for the target, linking in `armnnDeserializer` and `armnn` built for the target.
* Your program should call `armnnDeserializer:IDeserializer::Create` to get an IDeserializer, then call `IDeserializer::CreateNetworkFromBinary` on it, passing the contents of the .armnn file that you generated earlier.
* You will now have an armnn::INetworkPtr which you can pass to armnn::Optimize, then load into an instance of the Arm NN runtime for inference. These steps are the same as in the section "Optimize the model and load it onto a compute device" in https://developer.arm.com/solutions/machine-learning-on-arm/developer-material/how-to-guides/deploying-a-quantized-tensorflow-lite-mobilenet-v1-model-using-the-arm-nn-sdk/write-your-application

I hope that makes sense, please let me know if I can help further.

**Comment by Pradeepcbk** [comment]
@MatthewARM Thank you for the response.

I will try these steps with a simple model and get back to you if i end up in trouble.

I see that Arm Compute Library is not mentioned in any of your steps, is it not necessary in any of the steps or am i missing anything here?

Kindly let me know in case of any concerns.

Thanks in advance.

**Comment by MatthewARM** [comment]
Hi @Pradeepcbk In the step "Cross-compile your application for the target, linking in armnnDeserializer and armnn built for the target." you'll need to build Compute Library as a prerequisite for the Arm NN accelerated backends.

If you just need NEON CPU acceleration, you can just build the Compute Library NEON code (following the instructions in the Compute Library documentation). Enable the Arm NN "CpuAcc" backend via `-DARMCOMPUTENEON=1` passed to CMake. There is a How-To guide here: https://developer.arm.com/solutions/machine-learning-on-arm/developer-material/how-to-guides/configuring-the-arm-nn-sdk-build-environment-for-tensorflow-lite/before-you-begin that you can follow. This guide is intended to be for users of the Tensorflow Lite parser in Arm NN, but the steps about building Compute Library and Arm NN will be similar to what you need.

**Comment by Pradeepcbk** [comment]
@MatthewARM Thank you for the response.

In the link which you have shared, it says - Linux Distribution should be installed on the board. I have a customized OS installed as this is for a private organization. Is there a work around for this requirement? like a virtual distribution?

**Comment by akanxit** [comment]
@MatthewARM I need to use a pre trained tflite **lstm** model for inference on an arm board. Is it possible using armnn?  
Going through the comments, I understand I might need to convert the .tflite to .armnn and then use it for inference. Correct me if I am wrong.

**Comment by MikeJKelly** [comment]
Yes it is possible, providing the other operators in your model are supported. You will need to convert the .tflite file to an .armnn one using the process that Matthew has explained as the ArmNN TFLite parser does not support LSTM at this time.
